### ISSUE TITLE
CodeGenerator.findCommonStringDelimiter stack overflow

### ISSUE THREAD

**Issue by stefanpenner** [babel/babel#1963]
when printing objects with a large number of keys, example input:

https://gist.github.com/stefanpenner/7d3af870350d821bf9f2

```
RangeError: bower_components/simple-html-tokenizer/lib/simple-html-tokenizer/char-refs/full.js: Maximum call stack size exceeded
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
    at Function.findCommonStringDelimiter (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/generation/index.js:124:17)
```

**Comment by stefanpenner** [comment]
Should be relatively easy to port the code to not be recursive, I am just strapped for time this morning. I can likely implement later tonight.

If someone can beat me to it that would be lovely.

**Comment by stefanpenner** [comment]
Ah, `this` members are not yet included in the TCO optimization, and a member in class form can't refer to itself by name.

**Comment by sebmck** [comment]
Regression due to #1900.

**Comment by stefanpenner** [comment]
so interestingly, in a quick attempt to rework this as a pure function. TCO still isn't happening, although the same extraction does TCO in the repl http://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&code=function%20findCommonStringDelimiter(code%2C%20tokens%2C%20occurences%20%3D%20%7B%22'%22%3A%200%2C%20%22%5C%22%22%3A%200%7D)%20%7B%0A%20%20if%20(tokens.length%20%3D%3D%3D%200%20%7C%7C%20occurences%5B%22'%22%5D%20%2B%20occurences%5B%22%5C%22%22%5D%20%3E%3D%203)%20%7B%0A%20%20%20%20return%20occurences%5B%22'%22%5D%20%3E%20occurences%5B%22%5C%22%22%5D%20%3F%20%22single%22%20%3A%20%22double%22%3B%0A%20%20%7D%0A%0A%20%20if%20(tokens%5B0%5D.type.label%20%3D%3D%3D%20%22string%22)%20%7B%0A%20%20%20%20occurences%5Bcode%5Btokens%5B0%5D.start%5D%5D%2B%2B%3B%0A%20%20%7D%0A%0A%20%20return%20findCommonStringDelimiter(code%2C%20tokens.slice(1)%2C%20occurences)%3B%0A%7D

Is there some switch/option I am forgetting?

**Comment by sebmck** [comment]
I've reverted the commit via d335535f042b593a72cd39581aa729f07911ae2a, new release incoming. The TCO isn't performed because it doesn't know that `this.findCommonStringDelimiter` will always reference the same function.

**Comment by stefanpenner** [comment]
@sebmck in my extraction it is pure, no `this` involved. Still no TCO.

**Comment by stefanpenner** [comment]
example:

```
function findCommonStringDelimiter(code, tokens, occurences = {"'": 0, "\"": 0}) {
  if (tokens.length === 0 || occurences["'"] + occurences["\""] >= 3) {
    return occurences["'"] > occurences["\""] ? "single" : "double";
  }

  if (tokens[0].type.label === "string") {
    occurences[code[tokens[0].start]]++;
  }

  return findCommonStringDelimiter(code, tokens.slice(1), occurences);
}
```

**Comment by sebmck** [comment]
@stefanpenner Oh right, the recursive function approach isn't really beneficial and makes things less clear IMO so I think it's best to just stick to the older approach.

**Comment by stefanpenner** [comment]
@sebmck if I am doing something wrong with ^^ let me know. If not, and its a bug I'll gladly open an issue.

**Comment by stefanpenner** [comment]
missing piece: 

`"blacklist": ["es6.tailCall"],` in babel

**Comment by sebmck** [comment]
Fixed as of 5.6.17, thanks!
